knack/output.py

```diff
@@ -25,7 +25,7 @@ def _decode_str(output):
 class _ComplexEncoder(json.JSONEncoder):
 
     def default(self, o):  # pylint: disable=method-hidden
-        if isinstance(o, bytes) and not isinstance(o, str):
+        if (isinstance(o, bytes) or isinstance(o, bytearray)) and not isinstance(o, str):
```

Comment:
According to https://docs.python.org/3/library/stdtypes.html#bytes.decode

```py
bytes.decode(encoding='utf-8', errors='strict')
```

Knack's current implementation assumes `o` is bytes representing a string. This may not be true and will even cause error if `o` is a binary stream that can't be UTF-8 decoded.

It will be safer to use [`hex`](https://docs.python.org/3.10/library/stdtypes.html#bytearray.hex)/[base16](https://docs.python.org/3/library/base64.html#base64.b16encode) or [base64](https://docs.python.org/3/library/base64.html#base64.b64encode) to represent a binary stream.

I prefer not to change the current (incorrect) behavior of Knack. Instead, you may consider serializing the bytearray by yourself utilizing a transform. This will give your more control of how the bytearray should be serialized: 

https://github.com/Azure/azure-cli/blob/azure-cli-2.36.0/src/azure-cli/azure/cli/command_modules/role/commands.py#L28-L60

```py
def transform_graph_objects_with_cred(result):
...
    with self.command_group('ad app', client_factory=get_graph_client_applications, resource_type=PROFILE_TYPE,
                            exception_handler=graph_err_handler, transform=transform_graph_objects_with_cred) as g:
```

Reply:
Thanks for the suggestion about transform. I've opened another PR [#5575](https://github.com/Azure/azure-cli-extensions/pull/5575) to adopt this workaround.